issue #191 : google tts plugin is broken

olest opened the pull request:
The google api for TTS has changed and the current plugin is not longer working.

## Description
I have updated the plugin to the new google TTS api.


## Motivation and Context
As explained above, it changes the plugin to use the new api.

## How Has This Been Tested?
I don't know how to write tests for a tts plugin but I tried it using my google account and it works.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

olest committed:
issue #162 : festival plugin

olest committed:
Merge branch 'naomi-dev' of https://github.com/NaomiProject/Naomi into naomi-dev

olest committed:
Merge branch 'naomi-dev' of https://github.com/NaomiProject/Naomi into naomi-dev

olest committed:
Updating google stt plugin to work with new google api

olest committed:
Updating requirements

olest committed:
Adding unit test for google-stt plugin

olest committed:
Retrieve the location of the credentials file from the user's profile

aaronchantrill committed:
A few additional changes to google stt

I noticed that the test requires the environment variable to
exist in order to not be skipped, but then in the main google.py
module, you are getting the location of the file from profile.py.

I think this modification will favor the environment variable over
the profile.py setting (if both are set, the environment variable
wins) in both the google.py and tests/test_googlesttplugin.py.

I haven't had a chance to thoroughly test this yet, but I did
run unittest discovery against it with a valid authorization file
in profile.yml and that worked.

aaronchantrill committed:
Adding settings

Adding the settings structure that allows Naomi to query about
settings.

olest committed:
Merge pull request #1 from aaronchantrill/olest_google

A few additional changes to google stt

aaronchantrill committed:
Merge branch 'naomi-dev' into naomi-dev

olest committed:
Updating google-tts plugin. Small clarification to plugin.info for google-stt

olest committed:
google-tts: adding settings info

olest committed:
google-tts: clean up includes and comments

olest committed:
Resync with naomi-dev

olest committed:
Merge branch 'google-tts-update' into naomi-dev

aaronchantrill reviewed (changes requested):
First, of all, good job! You got text to speech working with the new Google Cloud!

Here is my experience so far.
First, due to an unrelated issue, I was unable to get the populate program to accept Google as my TTS engine. This was because there is a new variable which indicates that a voice has been chosen which never got set for the "Google" TTS choice. More on this later. We are currently in the process of transitioning away from populate.py to the "settings" system, so things are a bit mucked up at the moment.

After fixing populate.py to accept Google as a TTS engine, I got the following error:
```
WARNING:naomi.pluginstore:Plugin at '/home/pi/Naomi_GoogleTTS/plugins/tts/google-tts' skipped! (Reason: cannot import name 'texttospeech')
Traceback (most recent call last):
  File "/home/pi/Naomi_GoogleTTS/plugins/tts/google-tts/google.py", line 7, in <module>
    from google.cloud import texttospeech
ImportError: cannot import name 'texttospeech' 
```
To fix this error, I had to run the following command:
`pip3 install google-cloud-texttospeech`
without this, the plugin would not load, so never asked me the location of my credentials file.

After installing the texttospeech library, it did ask me the location of the credentials file which I entered. Then I got this message:
```
google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials.
Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run
the application. For more information, please see
https://cloud.google.com/docs/authentication/getting-started
```
I realized that although it is asking for a credentials file, it was not actually using the credentials file. I added the code in used for pulling the authentication information from the profile last time, and also changed the name of the setting to match what we ended up with in the Google STT engine (google: credentials_json:).

Now that the credentials were entered and set up, I got the following error message:
```
google.api_core.exceptions.PermissionDenied: 403 Cloud Text-to-Speech API has not been
used in project {project id} before or it is disabled. Enable it by visiting
https://console.developers.google.com/apis/api/texttospeech.googleapis.com/overview?project={project id} 
then retry. If you enabled this API recently, wait a few minutes for the action 
to propagate to our systems and retry.
```
Following the link in the message and signing into Google took me to the "Enable the Google Cloud Text to Speech" API button, so that was great. After clicking the button and waiting about 5 minutes, the Google TTS plugin started working.

While reading the documentation, I did find that there are a variety of different voices for different language choices. I started a python3 interactive session and entered the following lines:
```
>>> from google.cloud import texttospeech
>>> from google.oauth2 import service_account
>>> credentials_json = "/home/pi/.naomi/google_auth.json"
>>> cred = service_account.Credentials.from_service_account_file(credentials_json)
>>> client = texttospeech.TextToSpeechClient(credentials=cred)
>>> voices = client.list_voices()
>>> for voice in voices.voices:
...     if "en-US" in voice.language_codes:
...         print(voice)
... 
language_codes: "en-US"
name: "en-US-Wavenet-A"
ssml_gender: MALE
natural_sample_rate_hertz: 24000
...bunch more voices here...
language_codes: "en-US"
name: "en-US-Standard-E"
ssml_gender: FEMALE
natural_sample_rate_hertz: 24000
```
It would be nice to give the user a list of voices based on their selected language preference, then let them select and preview the voice they want to use.

At the same time, it would probably be good to let them know that the "WaveNet" voices are currently quite a bit more expensive than the "Standard" voices. From https://cloud.google.com/text-to-speech/pricing:

|Feature | Monthly free tier | Paid usage|
|--- | --- | ---|
|Standard (non-WaveNet) voices | 0 to 4 million characters | $4.00 USD / 1 million characters|
|WaveNet voices | 0 to 1 million characters | $16.00 USD / 1 million characters|

It's unlikely that a home user or developer would run up against those limits in a month, but it's good to keep in mind.

I am working on installation scripts for 3rd party developer plugins right now, including an install.py program that is run when the plugin is installed, updated, or first run, which might be a good way to ensure that the necessary python requirements get installed, and provide some additional information to the user during setup.

So, for now there are just a couple of changes that I had to make to get this working on my system. I'll do a pull request against your repository with those changes.

Thanks!

aaronchantrill committed:
A few fixes for Google-TTS

Fixed populate.py to accept "Google" as a TTS choice. Modified
the STT selection to use "google-tts" instead of "google" for
STT selection, as updated in the google stt plugin.info file.

Changed location of google authentication file setting to match
google-stt. It might make sense to change both of these to make
them use the proper root in profile.py:

google-tts:
  credentials_json

and

google-stt:
  credentials_json

rather than just

google:
  credentials_json

especially if someone wanted to use two different accounts for the
stt and tts functions.

Added code for importing and using the credentials file out of
profile.yml rather than setting the GOOGLE_APPLICATION_CREDENTIALS
environment variable.

olest committed:
Merge pull request #2 from aaronchantrill/Naomi-GoogleTTS

A few fixes for Google-TTS

aaronchantrill reviewed (approved):
Okay, looks good to me. I'm going to merge it.
